### ISSUE TITLE
[discussion] Composition vs inheritance

### ISSUE THREAD

**Issue by rosskevin** [dmtrKovalenko/material-ui-pickers#170]
I'm suggesting this in hopes of growing the community of contributors.

I'm a new contributor and I'm focused on tightening up the testing and definitions for typescript.  I'm finding it hard to follow because there is much duplication of properties plus there are component inheritance hierarchies which I have found to be very difficult to follow in react - and indeed I've rarely run into them in the wild.

I'm very strong in OO and I had a hard time thinking in react without inheritance, but once I did I saw the light of composition.  Discussed here: https://reactjs.org/docs/composition-vs-inheritance.html

Please consider switching to composition over inheritance for components as it is just another hurdle for a newcomer to understand the code, and this project sticks out as a rarity compared to most open source react.

Regardless, thanks for putting this work out there!

**Comment by cherniavskii** [comment]
@rosskevin thanks for sharing your concerns!
I absolutely agree with you, we need to get rid of inheritance in favor of composition!

**Comment by dmtrKovalenko** [comment]
@rosskevin  Hmm, thank you! I have thinking a lot about component inheritance and there I just "tried" to use that, but have not seen this article yet, thank I think we should refactor it.
